core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushLimitThroughUnion.java

```diff
@@ -73,8 +76,8 @@ public Result apply(LimitNode parent, Captures captures, Context context)
         ImmutableList.Builder<PlanNode> builder = ImmutableList.builder();
         boolean shouldApply = false;
         for (PlanNode source : unionNode.getSources()) {
-            // This check is to ensure that we don't fire the optimizer if it was previously applied.
-            if (isAtMost(source, context.getLookup(), parent.getCount())) {
+            // This check is to ensure that we don't fire the optimizer if it was previously applied, or if MergeUnion hasn't run yet
+            if (isNestedUnionSourceNode(source, context.getLookup()) || isAtMost(source, context.getLookup(), parent.getCount())) {
```

Comment:
I don't like this kind of special casing. They imply the order of application of rules matters and encode it in an out-of-band, implicit manner.

If I'm reading this correctly, what prevents the unions from being merged is the intervening partial limit:

```
- Limit
  - Union
    - Union
      - X
      - Y
    - Z

====>

- Limit (final)
  - Union
    - Limit (partial) <<<<<<<<<
      - Union
        - X
        - Y
    - Limit (partial)
      - Z
```

Alternatively, I'm thinking it might make sense to just push a partial limit through union without retaining the limit on top -- with the assumption and expectation that there's a final limit somewhere up the tree:

```
- Limit (final)
  - Union
    - Limit (partial) 
      - Union
        - X
        - Y
    - Limit (partial)
      - Z

====>

- Limit (final)
  - Union
      - Union
        - Limit (partial) 
          - X
        - Limit (partial) 
          - Y
    - Limit (partial)
      - Z
```

In that case, there would be no need to avoid pushing the limit through a union that contains nested unions.

Reply:
I think that approach implies two behaviors, the first to add partial limits underneath any and all unions, then another to specifically detect `limit -> union -> limit -> union` and specifically move the second limit underneath the second union. It's... a little strange looking and still feels like a special case- but it probably would work.

Without special case detection of the union / limit nesting scenario, you could remove a partial limit from on top of the union root which could be a problem for some plan shapes that don't actually have a final limit immediately on top of the union.

Reply:
> Without special case detection of the union / limit nesting scenario, you could remove a partial limit from on top of the union root which could be a problem for some plan shapes that don't actually have a final limit immediately on top of the union.

That shouldn't be an issue from a plan correctness perspective. The existence of a partial limit implies there's a final limit downstream somewhere.

> Importantly, if you did this in the two-phase approach, there's still a problem with rule ordering because the first and second phases can't co-exist in the same iterative optimizer without infinite-looping.

We'd need to change limit to have 3 styles: SINGLE, FINAL, PARTIAL. You'd only push down a limit below a union if it's SINGLE (by splitting into FINAL / PARTIAL) or PARTIAL. That would prevent the FINAL from being pushed down over and over.